### PR TITLE
infoschema,domain: introduce InfoSchemaV2, core data struct etc

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: ref #50959 

Problem Summary:

### What changed and how does it work?

This commit introduce the `infoschema_v2.go` file, include the core data struct, implement some of the APIs.

It's not a completed work, the code can barely run some basic operations.
There is still a long way to make the code smooth, but I fire this PR early and it is ready for review now. Why? 
Because I hope we can provide infoschema v2 in tidb 8.0, GA or at least RC, time is limited.
This commit provide a solid foundation that **we can work on together**. 

With the core data struct introduced, we can fullfill the missing things and split the tasks parallelly. 
Later, I, @GMHDBJD and other team members will continue the development.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Enable infoschema v2 here:

https://github.com/pingcap/tidb/compare/master...tiancaiamao:infoschema-v2-pr?expand=1#diff-d82e5bb6499a8d8c00f3d55ec95bccda3fab6d4cc45529cecf06552bfa973b1bR21

```
func init() {
    enableV2.Store(true)
}
```

Start tidb, create some tabe, insert data, drop table, verify the basic operations etc...

This commit doesn't break anyting because the new code is not used now.
We need prepare a test plan document and add tests later.

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
This is extracted from https://github.com/pingcap/tidb/pull/50591
The differ is that this pull request clean up the code and keep no breaking things, infoschema v1 is used.
In that branch, infoschema v2 is used and tested, and v1 does not load the data so we can see how much memory are used:
https://github.com/pingcap/tidb/blob/350b8876201c5fe55cfb773a4a02445dd691988c/pkg/infoschema/builder.go#L1140-L1144

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: pkg/infoschema/infoschema_v2.go [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+	// For the TableByName API, sorted by {dbName, tableName, tableID} => schemaVersion
+	//
+	// If the schema version +1 but a specific table does not change, the old record is
+	// kept and no new {dbName, tableName, tableID} => schemaVersion+1 record been added.
+	//
+	// It means as long as we can find a item in it, the item is available, even through the
+	// schema version maybe smaller than required.
+	//
+	// *IMPORTANT RESTRICTION*: Do we have the full data in memory? NO!
+	name2id *btree.BTreeG[Item]
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:72]: `byName`? or `tableItemsByName`

File: pkg/infoschema/infoschema_v2.go [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+	// Get from the cache.
+	key := cacheKey{id, is.schemaVersion}
+	tbl, found := is.cache.Get(key)
+	if found && tbl != nil {
+		return tbl, true
+	}
+
+	eq := func(a, b *Item) bool { return a.tableID == b.tableID }
+	itm, ok := search(is.byID, is.schemaVersion, Item{tableID: id, dbID: math.MaxInt64}, eq)
+	if !ok {
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:255]: if `byID` is not full data, shouldn't load from tikv?

File: pkg/infoschema/infoschema_v2.go [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+	ret, err := loadTableInfo(is.r, is.Data, id, itm.dbID, is.ts)
+	if err == nil {
+		is.cache.Set(key, ret)
+		return ret, true
+	}
+	return nil, false
+}
+
+func (is *infoschemaV2) TableByName(schema, tbl model.CIStr) (t table.Table, err error) {
+	if schema.L == "information_schema" || schema.L == "metrics_schema" || schema.L == "performance_schema" {
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:269]: use `InformationSchemaName` defined in util?

File: pkg/infoschema/infoschema_v2.go [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+}
+
+func (is *infoschemaV2) SchemaMetaVersion() int64 {
+	return is.schemaVersion
+}
+
+func (is *infoschemaV2) SchemaExists(schema model.CIStr) bool {
+	var ok bool
+	// TODO: support different version
+	is.schemaMap.Scan(func(item schemaItem) bool {
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:335]: use `Descend`?

File: pkg/infoschema/infoschema_v2.go [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+	// It provides the tableID => databaseID mapping.
+	//
+	// *IMPORTANT RESTRICTION*: Do we have the full data in memory? NO!
+	// But this mapping should be synced with name2id.
+	byID *btree.BTreeG[Item]
+
+	cache fifo.Cache[cacheKey, table.Table]
+
+	// For the SchemaByName API
+	schemaMap *btree.BTreeG[schemaItem]
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:85]: maybe add comments about how it's sorted as previous 2 fields do

File: pkg/infoschema/infoschema_v2.go [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+	// Using the KV timeout read feature to address the issue of potential DDL lease expiration when
+	// the meta region leader is slow.
+	snapshot.SetOption(kv.TiKVClientReadTimeout, uint64(3000)) // 3000ms.
+	m := meta.NewSnapshotMeta(snapshot)
+	tblInfos, err := m.ListSimpleTables(dbInfo.ID)
+	if err != nil {
+		if meta.ErrDBNotExists.Equal(err) {
+			return nil
+		}
+		panic(err)
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:383]: add a TODO here?

File: pkg/infoschema/infoschema_v2.go [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+	allocs := autoid.NewAllocatorsFromTblInfo(r, dbID, tblInfo)
+	b := NewBuilder(r, nil, infoData) // TODO: handle cached table!!!
+	ret, err := b.tableFromMeta(allocs, tblInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return ret, nil
+}
+
+var sf = &singleflight.Group{}
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:427]:
```suggestion
var loadTableSF = &singleflight.Group{}
```

File: pkg/infoschema/infoschema_v2.go [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+		// TODO load table panic!!!
+		panic(err)
+	}
+	tblInfo := res.(*model.TableInfo) // TODO: it could be missing!!!
+
+	ConvertCharsetCollateToLowerCaseIfNeed(tblInfo)
+	ConvertOldVersionUTF8ToUTF8MB4IfNeed(tblInfo)
+	allocs := autoid.NewAllocatorsFromTblInfo(r, dbID, tblInfo)
+	b := NewBuilder(r, nil, infoData) // TODO: handle cached table!!!
+	ret, err := b.tableFromMeta(allocs, tblInfo)
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:420]: odd to attach a util method to a `Builder` obj

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/infoschema/infoschema_v2.go (hunk 4 above)
Comment (tiancaiamao) [reply pkg/infoschema/infoschema_v2.go:335]: This is a full memory `Scan`, it does not matter whether we `Descend` or `Ascend`

File: pkg/infoschema/infoschema_v2.go (hunk 8 above)
Comment (tiancaiamao) [reply pkg/infoschema/infoschema_v2.go:420]: Well, cached table need some post operations ...

File: pkg/infoschema/infoschema_v2.go (hunk 5 above)
Comment (tiancaiamao) [reply pkg/infoschema/infoschema_v2.go:88]: Done

File: pkg/infoschema/infoschema_v2.go (hunk 6 above)
Comment (tiancaiamao) [reply pkg/infoschema/infoschema_v2.go:397]: Done

File: pkg/infoschema/infoschema_v2.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/infoschema/infoschema_v2.go:348]: seems `Descend` can only search items from [pivot, first], no need a full scan

**Review by D3Hunter — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]
LGTM for BR part

File: pkg/infoschema/infoschema_v2.go [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+			continue
+		}
+		tables = append(tables, tbl)
+	}
+	return
+}
+
+func loadTableInfo(r autoid.Requirement, infoData *Data, tblID, dbID int64, ts uint64) (table.Table, error) {
+	// Try to avoid repeated concurrency loading.
+	res, err, _ := loadTableSF.Do(fmt.Sprintf("%d-%d", dbID, tblID), func() (ret any, err error) {
```
Comment (tangenta) [inline pkg/infoschema/infoschema_v2.go:413]: Will there be concurrent queries using different `ts` on same `dbID-tblID`?

**Review by tangenta — APPROVED** [review]

**Pull request merged**